observer/observer.go

```diff
@@ -200,30 +222,35 @@ func (o *Observer) observe(ctx context.Context) error {
 }
 
 func (o *Observer) storeMessage(ctx context.Context, om *Message) error {
-	const insertMessage = `INSERT INTO latest_messages VALUES(?,?,?,?::json,?,?,?::json,?);`
-	voteMarshaled, err := json.Marshal(om.Vote)
-	if err != nil {
-		return fmt.Errorf("failed to marshal vote: %w", err)
-	}
-	var justificationMarshaled any
+	var justification any
 	if om.Justification != nil {
-		v, err := json.Marshal(om.Justification)
-		if err != nil {
-			return fmt.Errorf("failed to marshal justification: %w", err)
-		}
-		justificationMarshaled = string(v)
+		// Dereference to get the go-duckdb reflection in appender to behave. Otherwise,
+		// it fails as it interprets the type as mismatch due to how nullable fields are
+		// handled.
+		justification = *om.Justification
 	}
-	if _, err := o.db.ExecContext(ctx, insertMessage,
+	if err := o.dbAppender.AppendRow(
 		om.Timestamp,
 		om.NetworkName,
-		om.Sender,
-		string(voteMarshaled),
+		int64(om.Sender),
+		om.Vote,
 		om.Signature,
 		om.Ticket,
-		justificationMarshaled,
+		justification,
 		om.VoteValueKey,
 	); err != nil {
-		return fmt.Errorf("failed to execute query: %w", err)
+		return fmt.Errorf("failed to append row: %w", err)
+	}
+	o.unflushedMessageCount++
+	batchSizeReached := o.unflushedMessageCount >= o.maxBatchSize
+	batchDelayElapsed := !o.lastFlushedAt.IsZero() && time.Since(o.lastFlushedAt) >= o.maxBatchDelay
+	if batchSizeReached || batchDelayElapsed {
+		if err := o.dbAppender.Flush(); err != nil {
```

Comment:
Is `Flush` allowed to run concurrently?
I'm thinking, what if we run into low message count again, then the messages won't flush until another message lands.

Reply:
> Is Flush allowed to run concurrently?

I think it can't (godocs on that function are missing) but it is not being called concurrently in this implementation. 

> what if we run into low message count again, then the messages won't flush until another message lands
Yup good point. I'll adjust the channel to trigger flush on timer. This will be syncronous but would decouple flushing from arrival of messages. Thanks Kuba